### PR TITLE
feature(zc): add carracing in box2d

### PR DESCRIPTION
**Pull request by Super1ce**
## Description
Add CarRacing env and config of DQN
result:
image

## TODO

### CODE REVIEW

File: dizoo/box2d/carracing/config/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+from .carracing_dqn_config import carracing_dqn_config, carracing_dqn_create_config
```
Comment (puyuan1996) [inline dizoo/box2d/carracing/config/__init__.py:1]: 执行 bash format.sh dizoo/box2d/carracing 格式化文件

File: dizoo/box2d/carracing/envs/carracing_env.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    @property
+    def action_space(self) -> gym.spaces.Space:
+        return self._action_space
+
+    @property
+    def reward_space(self) -> gym.spaces.Space:
+        return self._reward_space
+
+    def __repr__(self) -> str:
+        return "DI-engine LunarLander Env"
```
Comment (puyuan1996) [inline dizoo/box2d/carracing/envs/carracing_env.py:161]: CarRacing

File: dizoo/box2d/carracing/envs/test_carracing_env.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+@pytest.mark.envtest
+@pytest.mark.parametrize(
+    'cfg', [
+        EasyDict({
+            'env_id': 'CarRacing-v1',
+            'act_scale': False
+        }),
+        EasyDict({
+            'env_id': 'CarRacingContinuous-v1',
```
Comment (puyuan1996) [inline dizoo/box2d/carracing/envs/test_carracing_env.py:15]: 为什么这里是v1, 其他文件是v2?

File: dizoo/box2d/carracing/envs/carracing_env.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+            )
+            self._init_flag = True
+        if hasattr(self, '_seed') and hasattr(self, '_dynamic_seed') and self._dynamic_seed:
+            np_seed = 100 * np.random.randint(1, 1000)
+            self._env.seed(self._seed + np_seed)
+        elif hasattr(self, '_seed'):
+            self._env.seed(self._seed)
+        self._eval_episode_return = 0
+        obs = self._env.reset()
+        obs = obs.astype(np.float32) / 255
```
Comment (puyuan1996) [inline dizoo/box2d/carracing/envs/carracing_env.py:78]: /255是必须的吗？如果不除会报错吗？或许可以加一个option选择是否除以255

File: dizoo/box2d/carracing/config/carracing_dqn_config.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+carracing_dqn_config = EasyDict(carracing_dqn_config)
+main_config = carracing_dqn_config
+
+carracing_dqn_create_config = dict(
+    env=dict(
+        type='carracing',
+        import_names=['dizoo.box2d.carracing.envs.carracing_env'],
+    ),
+    env_manager=dict(type='subprocess'),
+    # env_manager=dict(type='base'),
```
Comment (puyuan1996) [inline dizoo/box2d/carracing/config/carracing_dqn_config.py:72]: 没有用到的注释可以去掉

File: dizoo/box2d/carracing/config/carracing_dqn_config.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    policy=dict(type='dqn'),
+)
+carracing_dqn_create_config = EasyDict(carracing_dqn_create_config)
+create_config = carracing_dqn_create_config
+
+if __name__ == "__main__":
+    # or you can enter `ding -m serial -c lunarlander_dqn_config.py -s 0`
```
Comment (puyuan1996) [inline dizoo/box2d/carracing/config/carracing_dqn_config.py:79]: carraciing_dqn*

File: dizoo/box2d/carracing/config/carracing_dqn_config.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        continuous=False,
+        n_evaluator_episode=8,
+        stop_value=950,
+        # The path to save the game replay
+        # replay_path='./carracing_dqn_seed0/video',
+    ),
+    policy=dict(
+        # Whether to use cuda for network.
+        cuda=True,
+        load_path="./carracing_seed0/ckpt/ckpt_best.pth.tar",
```
Comment (puyuan1996) [inline dizoo/box2d/carracing/config/carracing_dqn_config.py:21]: load_path这里没有用到，可以删除了

File: dizoo/box2d/carracing/config/__init__.py (hunk 1 above)
Comment (puyuan1996) [reply dizoo/box2d/carracing/config/__init__.py:1]: commit的信息，也按照规范哈，类似这样 `feature(zc): add carring_env and related dqn config.`

File: dizoo/box2d/carracing/config/carracing_dqn_config.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
+from easydict import EasyDict
+
+nstep = 3
+carracing_dqn_config = dict(
+    exp_name='carracing_dqn_seed0',
+    env=dict(
+        # Whether to use shared memory. Only effective if "env_manager_type" is 'subprocess'
```
Comment (PaParaZz1) [inline dizoo/box2d/carracing/config/carracing_dqn_config.py:7]: why add these two line comments

File: dizoo/box2d/carracing/envs/carracing_env.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+import gym
+import numpy as np
+from easydict import EasyDict
+
+from ding.envs import BaseEnv, BaseEnvTimestep
+from ding.envs import ObsPlusPrevActRewWrapper
+from ding.envs.common import affine_transform, save_frames_as_gif
+from ding.torch_utils import to_ndarray
+from ding.utils import ENV_REGISTRY
+import cv2
```
Comment (PaParaZz1) [inline dizoo/box2d/carracing/envs/carracing_env.py:14]: 参考[文档](https://github.com/opendilab/DI-engine-docs/blob/main/source/21_code_style/index_zh.rst#%E5%90%88%E7%90%86%E7%9A%84-import-%E9%A1%BA%E5%BA%8F)修正import顺序

File: dizoo/box2d/carracing/envs/test_carracing_env.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+        assert env._seed == 314
+        obs = env.reset()
+        assert obs.shape == (96, 96, 3)
+        for i in range(10):
+            random_action = env.random_action()
+            timestep = env.step(random_action)
+            print(timestep)
+            assert isinstance(timestep.obs, np.ndarray)
+            assert isinstance(timestep.done, bool)
+            assert timestep.obs.shape == (96, 96, 3)
```
Comment (PaParaZz1) [inline dizoo/box2d/carracing/envs/test_carracing_env.py:37]: the shape should be (3, 96, 96)

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Comment by PaParaZz1** [comment]
add the information of this env in [env table](https://github.com/opendilab/DI-engine#environment-versatility)

File: dizoo/box2d/carracing/config/carracing_dqn_config.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
+from easydict import EasyDict
+
+nstep = 3
+carracing_dqn_config = dict(
+    exp_name='carracing_dqn_seed0-withouconfig',
```
Comment (PaParaZz1) [inline dizoo/box2d/carracing/config/carracing_dqn_config.py:5]: polish exp_name

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: dizoo/box2d/carracing/envs/carracing_env.py [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+    def __init__(self, cfg: dict) -> None:
+        self._cfg = cfg
+        self._init_flag = False
+        # env_id:CarRacing-v2
+        self._env_id = cfg.env_id
+        self._replay_path = None
+        self._replay_path_gif = cfg.replay_path_gif
+        self._save_replay_gif = cfg.save_replay_gif
+        self._save_replay_count = 0
+        if 'Continuous' in self._env_id:
```
Comment (PaParaZz1) [inline dizoo/box2d/carracing/envs/carracing_env.py:42]: remove `Continuous` related codes, which is abandoned in latest gym

File: dizoo/box2d/carracing/envs/carracing_env.py [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+                    self._env,
+                    video_folder=self._replay_path,
+                    episode_trigger=lambda episode_id: True,
+                    name_prefix='rl-video-{}'.format(id(self))
+                )
+            if hasattr(self._cfg, 'obs_plus_prev_action_reward') and self._cfg.obs_plus_prev_action_reward:
+                self._env = ObsPlusPrevActRewWrapper(self._env)
+            # self._observation_space = self._env.observation_space
+            self._observation_space = gym.spaces.Box(
+                low=np.min(self._env.observation_space.low),
```
Comment (PaParaZz1) [inline dizoo/box2d/carracing/envs/carracing_env.py:62]: correct range (after dividing 255)

File: dizoo/box2d/carracing/envs/carracing_env.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+        if not self._init_flag:
+            self._env = gym.make(self._cfg.env_id, continuous=self._cfg.continuous)
+            if self._replay_path is not None:
+                self._env = gym.wrappers.RecordVideo(
+                    self._env,
+                    video_folder=self._replay_path,
+                    episode_trigger=lambda episode_id: True,
+                    name_prefix='rl-video-{}'.format(id(self))
+                )
+            if hasattr(self._cfg, 'obs_plus_prev_action_reward') and self._cfg.obs_plus_prev_action_reward:
```
Comment (PaParaZz1) [inline dizoo/box2d/carracing/envs/carracing_env.py:58]: remove this wrapper, it is only used in some special cases.

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Review by PaParaZz1 — APPROVED** [review]

**Pull request merged**